stores/chain.go

```diff
@@ -0,0 +1,356 @@
+				BlockID:        hash256(index.ID),
+			}).Error; err != nil {
+			return nil
+		}
+	}
+
+	// remove outputs created at the reverted index
+	for _, e := range removed {
+		if err := u.tx.
+			Where("output_id", hash256(e.ID)).
+			Delete(&dbWalletOutput{}).
+			Error; err != nil {
+			return err
+		}
+	}
+
+	// remove events created at the reverted index
+	return u.tx.
+		Model(&dbWalletEvent{}).
+		Where("height = ? AND block_id = ?", index.Height, hash256(index.ID)).
+		Delete(&dbWalletEvent{}).
+		Error
+}
+
+// UpdateChainIndex updates the chain index in the database.
+func (u *chainUpdateTx) UpdateChainIndex(index types.ChainIndex) error {
+	return u.tx.
+		Model(&dbConsensusInfo{}).
+		Where(&dbConsensusInfo{Model: Model{ID: consensusInfoID}}).
+		Updates(map[string]interface{}{
+			"height":   index.Height,
+			"block_id": hash256(index.ID),
+		}).
+		Error
+}
+
+// UpdateContract updates the revision height, revision number, and size the
+// contract with given fcid.
+func (u *chainUpdateTx) UpdateContract(fcid types.FileContractID, revisionHeight, revisionNumber, size uint64) error {
+	updates := map[string]interface{}{
+		"revision_height": revisionHeight,
+	}
+	if revisionNumber > 0 {
+		updates["revision_number"] = fmt.Sprint(revisionNumber)
+	}
+	if size > 0 {
+		updates["size"] = size
+	}
+
+	if err := u.tx.
+		Model(&dbContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Updates(updates).
+		Error; err != nil {
+		return err
+	}
+	return u.tx.
+		Model(&dbArchivedContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Updates(updates).
+		Error
+}
+
+// UpdateContractState updates the state of the contract with given fcid.
+func (u *chainUpdateTx) UpdateContractState(fcid types.FileContractID, state api.ContractState) error {
+	var cs contractState
+	if err := cs.LoadString(string(state)); err != nil {
+		return err
+	}
+
+	if err := u.tx.
+		Model(&dbContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Update("state", cs).
+		Error; err != nil {
+		return err
+	}
+	return u.tx.
+		Model(&dbArchivedContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Update("state", cs).
+		Error
+}
+
+// UpdateContractProofHeight updates the proof height of the contract with given
+// fcid.
+func (u *chainUpdateTx) UpdateContractProofHeight(fcid types.FileContractID, proofHeight uint64) error {
+	if err := u.tx.
+		Model(&dbContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Update("proof_height", proofHeight).
+		Error; err != nil {
+		return err
+	}
+	return u.tx.
+		Model(&dbArchivedContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Update("proof_height", proofHeight).
+		Error
+}
+
+// UpdateFailedContracts marks active contract as failed if the current
+// blockheight surposses their window_end.
+func (u *chainUpdateTx) UpdateFailedContracts(blockHeight uint64) error {
+	return u.tx.
+		Model(&dbContract{}).
+		Where("state = ? AND ? > window_end", contractStateActive, blockHeight).
```

Comment:
(nit) you can split this up into 
```go
Where("state", contractStateActive).
Where("window_end", blockHeight).
```
since gorm will just merge them